### PR TITLE
refactor: migrate infra as code to terraform

### PR DESCRIPTION
**Pull request by lucacasonato**
Terraform is a lot easier to use, and has integrations with multiple providers (AWS, MongoDB, Cloudflare, ect). Ultimatly we can deploy all of the required infra through Terraform.

### CODE REVIEW

**Comment by wperron** [comment]
my TODO list before we merge:
- [ ] tag all the things! (mostly for better observability in the bill)
- [ ] add CORS config to the storage bucket (see the SAM one for reference)
- [ ] add more granular ACL to the storage bucket to match the existing bucket
- [ ] ideally create a script to sync up the storage buckets
- [ ] test s3 -> s3 sync speeds (no code required)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -15,48 +15,66 @@ jobs:
       - name: Setup Deno environment
         uses: denolib/setup-deno@v1.3.0
         with:
-          deno-version: v1.2.2
+          deno-version: v1.3.0
```
Comment (ry) [inline .github/workflows/ci.yml:18]: not 1.3.2?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -15,48 +15,66 @@ jobs:
...
+            -var domain=$DOMAIN \
+            -var certificate_arn=$CERTIFICATE_ARN \
+            -out plan.tfplan
         env:
-          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
-          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          MONGODB_URI: ${{ secrets.MONGODB_URI }}
+          DOMAIN: staging.api.deno.land
+          CERTIFICATE_ARN: arn:aws:acm:us-east-1:035475582903:certificate/1c3bd085-95ef-4c27-ae09-18d27825739c
+          TERRAFORM_STATE_BUCKET: deno-registry2-prod-terraform-state-b3a31d16
```
Comment (ry) [inline .github/workflows/ci.yml:59]: Where does the b3a31d16 identifier come from?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (lucacasonato) [reply .github/workflows/ci.yml:18]: I tested with 1.3.0, will update to 1.3.2 later, once I tested that.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (lucacasonato) [reply .github/workflows/ci.yml:59]: This is a randomly generated suffix from when this bucket was created.

File: terraform/builds.tf [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  connection_type        = "INTERNET"
+  integration_uri        = aws_lambda_function.builds_get.invoke_arn
+  payload_format_version = "2.0"
+}
+
+resource "aws_apigatewayv2_route" "builds_get" {
+  api_id    = aws_apigatewayv2_api.deno_api.id
+  route_key = "GET /builds/{id}"
+  target    = "integrations/${aws_apigatewayv2_integration.builds_get.id}"
+}
```
Comment (ry) [inline terraform/builds.tf:53]: this is my first introduction to terraform .. looks pretty readable / reasonable

**Pull request merged**